qiskit/algorithms/aux_ops_evaluator.py

```diff
@@ -0,0 +1,176 @@
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""Evaluator of auxiliary operators for algorithms."""
+
+from typing import Tuple, Union, List
+
+import numpy as np
+
+from qiskit import QuantumCircuit
+from qiskit.algorithms.minimum_eigen_solvers.minimum_eigen_solver import ListOrDict
+from qiskit.opflow import (
+    CircuitSampler,
+    ListOp,
+    StateFn,
+    OperatorBase,
+    ExpectationBase,
+)
+from qiskit.providers import BaseBackend, Backend
+from qiskit.quantum_info import Statevector
+from qiskit.utils import QuantumInstance
+
+
+def eval_observables(
+    quantum_instance: Union[QuantumInstance, BaseBackend, Backend],
+    quantum_state: Union[
+        Statevector,
+        QuantumCircuit,
+        OperatorBase,
+    ],
+    observables: ListOrDict[OperatorBase],
+    expectation: ExpectationBase,
+    threshold: float = 1e-12,
+) -> ListOrDict[Tuple[complex, complex]]:
+    """
+    Accepts a list or a dictionary of operators and calculates their expectation values - means
+    and standard deviations. They are calculated with respect to a quantum state provided. A user
+    can optionally provide a threshold value which filters mean values falling below the threshold.
+
+    Args:
+        quantum_instance: A quantum instance used for calculations.
+        quantum_state: An unparametrized quantum circuit representing a quantum state that
+            expectation values are computed against.
+        observables: A list or a dictionary of operators whose expectation values are to be
+            calculated.
+        expectation: An instance of ExpectationBase which defines a method for calculating
+            expectation values.
+        threshold: A threshold value that defines which mean values should be neglected (helpful for
+            ignoring numerical instabilities close to 0).
+
+    Returns:
+        A list or a dictionary of tuples (mean, standard deviation).
+
+    Raises:
+        ValueError: If a ``quantum_state`` with free parameters is provided.
+    """
+
+    if (
+        isinstance(
+            quantum_state, (QuantumCircuit, OperatorBase)
+        )  # Statevector cannot be parametrized
+        and len(quantum_state.parameters) > 0
+    ):
+        raise ValueError(
+            "A parametrized representation of a quantum_state was provided. It is not "
+            "allowed - it cannot have free parameters."
+        )
+
+    # Create new CircuitSampler to avoid breaking existing one's caches.
+    sampler = CircuitSampler(quantum_instance)
+
+    list_op = _prepare_list_op(observables)
+
+    observables_expect = expectation.convert(
+        StateFn(list_op, is_measurement=True).compose(StateFn(quantum_state))
+    )
```

Comment:
The `PauliExpectation` (and the others likely too?) cannot handle  a ListOp that is a measurement. That's certainly a case we should support, but the following works:
```suggestion
    expectation_values = ListOp([StateFn(op, is_measurement=True).compose(StateFn(quantum_state)) for op in observables])
    expectation_converted = expectation.convert(expectation_values)
```